Store last time index as column on DataFrame

tamargrey opened the pull request:
Stores `last_time_index` as a column in the Woodwork dataframe.

Identifying factors of the last time index column:
- Name of the column stored in `df.ww.metadata` under `'last_time_index'`
- The column will be names `last_time` (unless there's overlap between names - tbd what we'll do there)
- The semantic tag `last_time_index` will be added to the column in the woodwork table 

Some of these (the semantic tag) might be overkill, but I figured more things that indicate that the column is additional data the better.

To Do:

- [x] Determine why we started having to add `astype('object')` before converting to datetime (possibly related to the dtype being different somehow when the lti is coming off a woodwork dataframe??)
- [x] Unexplained bugs for `add_last_time_indexes` when called from update_dataframe
  - [x] In `test_update_dataframe_and_lti` find way to test koalas with timedelta and/or mismatched index
  - [x] `test_update_dataframe_last_time_index` koalas and dask failures
- [x] Fix koalas inability to reorder indices (check calculate feature matrix to see if there's something there)
- [x] Additional woodwork testing for last time index in `test_ww_es`
- [x] Update `__sizeof__` to no longer look at metadata for the last time index
- [x] confirm `test_to_parquet_with_lti` and `test_operations_invalidate_metadata` are working correctly (might have to wait till after) #1452 
- [x] confirm bug with not being able to deepcopy koalas dataframes is gone

 
closes #1254

tamargrey commented on featuretools/entityset/entityset.py:
```diff
@@ -1034,14 +1044,20 @@ def add_last_time_indexes(self, updated_dataframes=None):
+        for df in self.dataframes:
+            lti = es_lti_dict[df.ww.name]
+            if lti is not None:
+                df.ww['last_time'] = lti
```
This introduces nans any time the underlying index of the dataframe doesn't match the underlying index of `lti`. For pandas, this shouldn't be an issue because it handles differently ordered indexes well, and the underlying index will always match the woodwork index.

But because dask and koalas never set the underlying index of their dataframes, this means that unless the indexes happen to match, there will be nans. (this is most easily seen with `es.set_last_time_index(['products'])` since products has a categorical index. 

I tried mapping `lti`'s underlying index to the woodwork index, but that proved to be messy, and I'm not entirely sure possible with dask. I also tried using just the underlying index throughout last time index calculation, but because of the `groupby` on the woodwork index values, we'd still need a way of mapping to the actual underlying index.

tamargrey replied:
Solved by using `merge` to combine the dataframe with the last time index. This creates a new DataFrame object for Dask and Koalas. Pandas still uses the setitem, so the object is the same.

tamargrey commented on featuretools/entityset/entityset.py:
```diff
@@ -975,25 +982,28 @@ def add_last_time_indexes(self, updated_dataframes=None):
                     for df in child_dataframes:
                         if df.ww.name not in explored and df.ww.name not in [q.ww.name for q in queue]:
+                            # must also reset last time index here
+                            es_lti_dict[df.ww.name] = None
```
This allows us to reset the children that don't get added to the queue until now (whereas the parents get added at the beginning, and we set the last time indexes to be None there)

tamargrey commented on featuretools/entityset/entityset.py:
```diff
@@ -1034,14 +1044,74 @@ def add_last_time_indexes(self, updated_dataframes=None):
+                    # --> need to make sure this works for dask and koalas
+                    if lti.dtype == 'datetime64[ns]':
+                        lti = lti.apply(lambda x: x.value)
+                    lti = ww.init_series(lti, logical_type='Double')
```
I think we've always stored the last time index as Datetimes, but now that we're putting it into the DataFrame, I was thinking it should match the entityset's time type. Happy to remove these lines if it's not necessary, though.

tamargrey commented on featuretools/entityset/entityset.py:
```diff
@@ -1034,14 +1044,74 @@ def add_last_time_indexes(self, updated_dataframes=None):
+                    )
+
+                    new_idx = new_df[new_df.ww.index]
+                    new_idx.name = None
```
Added these lines to set the underlying index to match the Woodwork index for Dask because the underlying index was being completely changed by the merge operation (I think it has something to do with the partitions), but it seemed better to have the index match the Woodwork index rather than have a non-unique index.

tamargrey commented on featuretools/entityset/entityset.py:
```diff
@@ -1034,14 +1044,74 @@ def add_last_time_indexes(self, updated_dataframes=None):
+                        semantic_tags={col_name: tags - {'index', 'time_index'} for col_name, tags in df.ww.semantic_tags.items()},
+                        table_metadata=df.ww.metadata,
+                        column_metadata={col_name: col_schema.metadata for col_name, col_schema in df.ww.columns.items()},
+                        use_standard_tags=df.ww.use_standard_tags
```
I'm not the biggest fan of this way of initializing Woodwork. It seems very likely to cause bugs/need updating every time Woodwork accessors get updated. 

But we can't initialize with a TableSchema here because the merge operations often change the dtypes, so the schema wouldn't be valid even after we went through the trouble of updating it to contain the last time index column. I played around with the idea of a function to allow initializing Woodwork with a partial schema (or maybe a partial columns dictionary?) while performing type inference on the columns without any schemas. I think this would also be useful for Woodwork's concat.

thehomebrewnerd replied:
> I played around with the idea of a function to allow initializing Woodwork with a partial schema

I like the idea of this type of behavior - use as much as we can and infer the rest. I think this might be useful in other situations beyond concat as well. Maybe we should add a couple issues - one to add this capability to Woodwork and a second to circle back and clean this up later?

tamargrey replied:
Sounds good, I'll create the issues!

thehomebrewnerd reviewed (commented):
@tamargrey Just a couple quick comments on this and I'll do a deeper review later today.

I think there are two tests that still need to be fixed. The tests are `test_add_last_time_indexes` in `test_dask_es.py` and `test_koalas_es.py`.

Also, I created a test branch and merged in this code with the serialization update code from my PR and confirmed that `test_to_parquet_with_lti` and `test_operations_invalidate_metadata` both pass with the combined code. I went ahead and marked that item complete in the to-do list.

tamargrey commented:
@thehomebrewnerd thanks for checking the serialization stuff! Updated the koalas and dask tests and also realized that our `int_es` fixture was only testing pandas, so I updated it to also use dask and koalas in conftest. Confirmed that the other tests that use `int_es` and were running before were still running, but if if doesn't make sense to change that fixture because of the other tests that use it, happy to just have them exist for the woodwork tests.

thehomebrewnerd reviewed (commented):
Not looking too bad. I still want to think through the logic inside `add_last_time_indexes` a bit more, but in the meantime here is a bit more feedback.

thehomebrewnerd commented on featuretools/tests/entityset_tests/test_ww_es.py:
```diff
@@ -624,23 +679,62 @@ def test_update_dataframe_different_dataframe_types():
+    new_dataframe = es['log'].copy()
+    new_dataframe['datetime'] = new_time_index
+
+    es.update_dataframe('log', new_dataframe, recalculate_last_time_indexes=True)
```
This might also be an issue for the tests that follow this one.

tamargrey replied:
Oof, yeah, this definitely is something we need to handle. Good catch--added. 

In the case where we aren't recalculating last time index and the new dataframe has no last time index column, I opted to lose any reference to a last time index on the DataFrame, since the new dataframe won't have a last time index column added.

thehomebrewnerd replied:
> In the case where we aren't recalculating last time index and the new dataframe has no last time index column, I opted to lose any reference to a last time index on the DataFrame, since the new dataframe won't have a last time index column added.

Seems reasonable

tamargrey commented on featuretools/tests/entityset_tests/test_ww_es.py:
```diff
@@ -286,11 +287,65 @@ def test_update_dataframe():
+    assert isinstance(es['products'].ww.logical_types['last_time'], Datetime)
+
+
+def test_add_last_time_non_numeric_index(pd_es, ks_es, dask_es):
+    pd_es.add_last_time_indexes(['products'])
+    dask_es.add_last_time_indexes(['products'])
+    ks_es.add_last_time_indexes(['products'])
+
+    assert list(to_pandas(pd_es['products']['last_time']).sort_index()) == list(to_pandas(dask_es['products']['last_time']).sort_index())
+    assert list(to_pandas(pd_es['products']['last_time']).sort_index()) == list(to_pandas(ks_es['products']).sort_values('id')['last_time'])
+
+    assert pd_es['products'].ww.schema == dask_es['products'].ww.schema
+    assert pd_es['products'].ww.schema == ks_es['products'].ww.schema
```
Because `products` is the only dataframe in the mock entityset that doesn't have a numeric index, there were some bugs that were only showing up when the koalas and dask underlying indices (which will be numeric and therefore often happened to exactly match the woodwork index column) couldn't possibly match. 

So this test is confirming that the results of add_last_time index actually match in that case between pandas, dask, and koalas. It also confirms that the schemas match between the cases, because we're reinitializing Woodwork on a new DataFrame for the dask and koalas cases, and I wanted to make sure that was maintained as well.

thehomebrewnerd replied:
Got it. Thanks. Maybe a short code comment in the test would be helpful to help us remember that detail?

tamargrey replied:
added a comment!

tamargrey commented on featuretools/tests/entityset_tests/test_ww_es.py:
```diff
@@ -286,11 +287,65 @@ def test_update_dataframe():
+
+    es.add_last_time_indexes(['customers'])
+
+    # Current behavior will attempt to overwrite this column.
```
That's a good point about the uniqueness of the name. My original thought processes around overwriting the column was to make sure that a user could call `add_last_time_indexes` twice without issues, but we probably want to minimize the chance that we overwrite an actual column. 

options to consider:
- `ft_last_time`
- `last_time_in_entityset`

Another option to avoid overwriting a user's column is to confirm that the column contains the `last_time_index` tag or that it's listed in the metadata as the last time index before proceeding. And raising an error telling users to rename otherwise

thehomebrewnerd replied:
Yeah, I think it makes sense to allow overwriting a column that featuretools has set. I like the idea of using the semantic tag for this purpose - couple that with a unique column name and I think that should make it highly unlikely that we would overwrite user data.

Could also do something like add an underscore to the column name to make it less likely to be something already present up `_ft_last_time`, for example.

tamargrey replied:
Added the column name change and semantic tag check

thehomebrewnerd commented on featuretools/entityset/entityset.py:
```diff
@@ -898,11 +898,13 @@ def add_last_time_indexes(self, updated_dataframes=None):
             updated_dataframes (list[str]): List of dataframe names to update last_time_index for
                 (will update all parents of those dataframes as well)
         """
+        LTI_COLUMN_NAME = '_ft_last_time'
```
I like this change - easy to update in the future if we need to! What do you think about moving this out of the method and also using this same variable in our test assertions?

thehomebrewnerd commented on featuretools/tests/entityset_tests/test_ww_es.py:
```diff
@@ -715,12 +735,32 @@ def test_update_dataframe_dont_recalculate_last_time_index(es):
 
 
-def test_update_dataframe_recalculate_last_time_index(es):
+def test_update_dataframe_recalcuate_last_time_index_present(es):
```
`recalcuate` -> `recalculate`

thehomebrewnerd replied:
Also, if I'm following this test correctly, the original last time index for the log entity is the same as the recalculated last time index. I'm thinking we might want to do something here that causes the last time index after `update_dataframe` to change from what it was before - otherwise I'm not sure we can actually verify that it was recalculated as opposed to just carried over from before.

tamargrey replied:
Yep, good call. Added a line where we confirm that the last time index column also matches the `new_time_index` here and in the test above this (which is the same except for whether or not the lti column is on the new dataframe).

thehomebrewnerd commented on featuretools/tests/entityset_tests/test_ww_es.py:
```diff
@@ -715,12 +735,32 @@ def test_update_dataframe_dont_recalculate_last_time_index(es):
+    new_dataframe.pop('_ft_last_time')
+
+    es.update_dataframe('log', new_dataframe, recalculate_last_time_indexes=True)
+    pd.testing.assert_series_equal(to_pandas(es['log']['datetime']).sort_index(), to_pandas(new_time_index).sort_index(), check_names=False)
```
Do we also need to verify that the last time index column is now present after the update?

tamargrey replied:
yep--the check for whether the lti column matches what we expect accesses `es['log'][LTI_COLUMN_NAME]` so that should cover it!

thehomebrewnerd commented on featuretools/entityset/api.py:
```diff
@@ -1,6 +1,6 @@
 from .deserialize import read_entityset
 from .entity import Entity
-from .entityset import EntitySet
+from .entityset import LTI_COLUMN_NAME, EntitySet
```
Does this need to be here?

tamargrey replied:
probably not; I was trying to avoid two imports for `from featuretools.entityset import LTI_COLUMN_NAME, EntitySet` in `test_ww_es` but can just change it to `from featuretools.entityset.entityset import LTI_COLUMN_NAME, EntitySet`

thehomebrewnerd reviewed (approved):
Looking good. Only one minor comment this time around.
